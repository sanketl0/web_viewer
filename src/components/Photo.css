/* Smaller loading text */
.psv-loader-text {
	text-align: center;
	font-size: 0.9em;
}

.psv-loader-image { width: 80%; }

.psv-overlay { opacity: 1 !important; }

/* Hide PSV tooltips */
.psv-virtual-tour-tooltip { display: none; }

/* Resize canvas for print */
@media print {
	.gvs .gvs-psv canvas {
		width: 100% !important;
		height: unset !important;
	}
}

/* Virtual Tour SVG arrows */
.gvs-psv-tour-arrows {
	background: none;
	border: none;

}
.gvs-psv-tour-arrows:active svg,
.gvs-psv-tour-arrows.gvs-clicked svg {
	opacity: 0.4;
	transition: opacity 0.2s ease;
}
.gvs-psv-tour-arrows svg {
	width: 100%;
	height: auto;
}

/* No virtual tour arrows if photo is reduced */
.gvs-mini .gvs-psv .psv-virtual-tour-arrows {
	display: none;
}