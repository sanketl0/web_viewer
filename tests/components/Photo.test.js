import Photo from "../../src/components/Photo";
import fs from "fs";
import path from "path";

jest.mock("@photo-sphere-viewer/core", () => ({
	Viewer: class {
		constructor(opts) {
			this._psvOpts = opts;
			this.loader = {
				canvas: { setAttribute: jest.fn() },
				__updateContent: jest.fn(),
				show: jest.fn(),
			};
			this.renderer = {
				renderer: {
					toneMapping: null,
					toneMappingExposure: null,
				}
			};
		}
		addEventListener() {;}
		getPlugin() {
			return {
				addEventListener: jest.fn(),
				datasource: {
					nodeResolver: jest.fn(),
				},
				arrowsRenderer: {
					clear: jest.fn(),
				},
				state: {
					currentNode: null,
					datasource: { nodes: {} },
				},
				config: {
					transitionOptions: jest.fn(),
				},
				__onEnterObject: jest.fn(),
				__onLeaveObject: jest.fn(),
			};
		}
	}
}));

jest.mock("@photo-sphere-viewer/equirectangular-tiles-adapter", () => ({
	EquirectangularTilesAdapter: jest.fn(),
}));

jest.mock("@photo-sphere-viewer/virtual-tour-plugin", () => ({
	VirtualTourPlugin: jest.fn(),
}));

const createParent = () => ({
	addEventListener: jest.fn(),
	dispatchEvent: jest.fn(),
	isWidthSmall: jest.fn(),
	select: jest.fn(),
	_t: { gvs: {} },
	_api: {
		_getFetchOptions: jest.fn(),
		getPictureMetadataUrl: (picId, seqId) => `https://geovisio.fr/api/collections/${seqId}/items/${picId}`
	},
});


describe("constructor", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		expect(c.className).toBe("gvs-psv");
	});
});

describe("_getNodeFromAPI", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("works", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		global.fetch = jest.fn(() =>
			Promise.resolve({
				ok: true,
				json: () => Promise.resolve(JSON.parse(fs.readFileSync(path.join(__dirname, "..", "data", "Viewer_pictures_1.json")))),
			})
		);
		global.Date = jest.fn(() => ({ toLocaleDateString: () => "June 3 2022" }));
		const res = await ph._getNodeFromAPI("id");
		expect(res).toMatchSnapshot();
	});

	it("works with nav filter", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		p._picturesNavFilter = () => false;
		global.fetch = jest.fn(() =>
			Promise.resolve({
				ok: true,
				json: () => Promise.resolve(JSON.parse(fs.readFileSync(path.join(__dirname, "..", "data", "Viewer_pictures_1.json")))),
			})
		);
		global.Date = jest.fn(() => ({ toLocaleDateString: () => "June 3 2022" }));
		const res = await ph._getNodeFromAPI("id");
		expect(res).toMatchSnapshot();
	});
});

describe("getPictureMetadata", () => {
	it("works when pic is selected", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour.state.currentNode = { bla: "bla" };
		expect(ph.getPictureMetadata()).toStrictEqual({ bla: "bla" });
	});

	it("nulls when no pic is selected", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		expect(ph.getPictureMetadata()).toBeNull();
	});
});

describe("_onSelect", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { setCurrentNode: jest.fn(() => Promise.resolve()), getCurrentNode: jest.fn() };
		ph._onSelect({ detail: { picId: "id" } });
		expect(ph._myVTour.setCurrentNode.mock.calls).toEqual([["id"]]);
	});

	it("works on pic ID already used", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { setCurrentNode: jest.fn(() => Promise.resolve()), getCurrentNode: () => "id" };
		ph._onSelect({ detail: { picId: "id" } });
		expect(ph._myVTour.setCurrentNode.mock.calls).toEqual([["id"]]);
	});
});

describe("goToNextPicture", () => {
	it("fails if no current picture", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { state: { currentNode: undefined } };
		expect(() => ph.goToNextPicture()).toThrow(new Error("No picture currently selected"));
	});

	it("works if next pic exists", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { state: { currentNode: { sequence: { id: "seq", nextPic: "idnext" } } } };
		ph.goToNextPicture();
		expect(p.select.mock.calls).toEqual([["seq", "idnext"]]);
	});

	it("fails if no next picture", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { state: { currentNode: { sequence: {} } } };
		expect(() => ph.goToNextPicture()).toThrow(new Error("No next picture available"));
	});
});

describe("goToPrevPicture", () => {
	it("fails if no current picture", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { state: { currentNode: undefined } };
		expect(() => ph.goToPrevPicture()).toThrow(new Error("No picture currently selected"));
	});

	it("works if next pic exists", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { state: { currentNode: { sequence: { id: "seq", prevPic: "idprev" } } } };
		ph.goToPrevPicture();
		expect(p.select.mock.calls).toEqual([["seq", "idprev"]]);
	});

	it("fails if no next picture", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour = { state: { currentNode: { sequence: {} } } };
		expect(() => ph.goToPrevPicture()).toThrow(new Error("No previous picture available"));
	});
});

describe("goToPosition", () => {
	it("works", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		p._api = {
			getPicturesAroundCoordinates: () => Promise.resolve(JSON.parse(fs.readFileSync(path.join(__dirname, "..", "data", "Viewer_pictures_1.json")))),
			_getFetchOptions: jest.fn()
		};

		const res = await ph.goToPosition(48.7, -1.8);

		expect(res).toEqual("0005086d-65eb-4a90-9764-86b3661aaa77");
		expect(p.select.mock.calls).toEqual([["bb129602-5ac1-4512-bf67-9ec1fa23033f", "0005086d-65eb-4a90-9764-86b3661aaa77"]]);
	});

	it("handles empty result from API", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		p._api = {
			getPicturesAroundCoordinates: () => Promise.resolve({ "features": [] }),
			_getFetchOptions: jest.fn()
		};
		return expect(ph.goToPosition()).rejects.toStrictEqual(new Error("No picture found nearby given coordinates"));
	});
});

describe("getXY", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph.getPosition = () => ({ yaw: 0.7853981634, pitch: -1.2217304764 });
		expect(ph.getXY()).toEqual({ x: 45.0000000001462, y: -70.00000000022743 });
	});
});

describe("getXYZ", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph.getPosition = () => ({ yaw: 0.7853981634, pitch: -1.2217304764 });
		ph.getZoomLevel = () => 15;
		expect(ph.getXYZ()).toEqual({ x: 45.0000000001462, y: -70.00000000022743, z: 15 });
	});
});

describe("clearPictureMetadataCache", () => {
	it("works when no pic is selected", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		await ph.clearPictureMetadataCache();
		expect(ph._myVTour.state.currentNode).toBeNull();
		expect(ph._myVTour.state.datasource.nodes).toStrictEqual({});
	});

	it("works when a pic is selected", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph._myVTour.state.currentNode = { id: "pic", sequence: { id: "seq" } };
		await ph.clearPictureMetadataCache();
		expect(p.select.mock.calls).toMatchSnapshot();
	});
});

describe("setXYZ", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph.rotate = jest.fn();
		ph.zoom = jest.fn();
		ph.setXYZ(45, -45, 3);
		expect(ph.zoom.mock.calls).toEqual([[3]]);
		expect(ph.rotate.mock.calls).toEqual([[{ yaw: 0.7853981633974483, pitch: -0.7853981633974483 }]]);
	});
});

describe("setHigherContrast", () => {
	it("works on enable", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph.needsUpdate = jest.fn();
		ph.setHigherContrast(true);
		expect(ph.renderer.renderer.toneMapping).toBe(3);
		expect(ph.renderer.renderer.toneMappingExposure).toBe(2);
		expect(ph.needsUpdate.mock.calls.length).toBe(1);
	});

	it("works on disable", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph.needsUpdate = jest.fn();
		ph.setHigherContrast(false);
		expect(ph.renderer.renderer.toneMapping).toBe(0);
		expect(ph.renderer.renderer.toneMappingExposure).toBe(1);
		expect(ph.needsUpdate.mock.calls.length).toBe(1);
	});
});

describe("getTransitionDuration", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c, { transitionDuration: 42 });
		expect(ph.getTransitionDuration()).toBe(42);
	});
});

describe("setTransitionDuration", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		ph.setTransitionDuration(1024);
		expect(ph.getTransitionDuration()).toBe(1024);
		expect(p.dispatchEvent.mock.calls).toMatchSnapshot();
	});

	it("fails when value is invalid", () => {
		const p = createParent();
		const c = document.createElement("div");
		const ph = new Photo(p, c);
		expect(() => ph.setTransitionDuration(-1)).toThrowError("Invalid transition duration (should be between 100 and 3000)");
		expect(() => ph.setTransitionDuration(3001)).toThrowError("Invalid transition duration (should be between 100 and 3000)");
	});
});
