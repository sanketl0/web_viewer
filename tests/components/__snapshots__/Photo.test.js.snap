// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_getNodeFromAPI works 1`] = `
Object {
  "caption": Object {
    "date": Object {
      "toLocaleDateString": [Function],
    },
    "license": "proprietary",
    "producer": "User",
  },
  "gps": Array [
    2.1822941,
    48.9810589,
  ],
  "horizontalFov": 360,
  "id": "0005086d-65eb-4a90-9764-86b3661aaa77",
  "links": Array [
    Object {
      "arrowStyle": Object {
        "element": <img
          alt=""
          src="arrow_triangle.svg"
        />,
        "size": Object {
          "height": 75,
          "width": 75,
        },
      },
      "gps": Array [
        2.1822633,
        48.9811134,
      ],
      "linkOffset": Object {
        "yaw": -0,
      },
      "nodeId": "ccf5f468-38b8-4b42-8d79-b97f5391e0a3",
    },
    Object {
      "arrowStyle": Object {
        "element": <img
          alt=""
          src="arrow_triangle.svg"
        />,
        "size": Object {
          "height": 75,
          "width": 75,
        },
      },
      "gps": Array [
        2.1823272,
        48.9810024,
      ],
      "linkOffset": Object {
        "yaw": -0,
      },
      "nodeId": "127d9282-d082-46ca-bc09-4b0d8569dd2c",
    },
  ],
  "panorama": Object {
    "basePanoData": [Function],
    "baseUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/sd.webp",
    "cols": 8,
    "hdUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/hd.jpg",
    "origBaseUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/sd.webp",
    "rows": 4,
    "thumbUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/thumb.jpg",
    "tileUrl": [Function],
    "width": 5760,
  },
  "properties": Object {
    "datetime": "2022-05-03T13:28:56+00:00",
    "license": "proprietary",
    "pers:interior_orientation": Object {
      "camera_manufacturer": "GoPro",
      "camera_model": "Max",
      "field_of_view": 360,
      "focal_length": 3,
    },
    "tiles:tile_matrix_sets": Object {
      "geovisio": Object {
        "identifier": "geovisio-0005086d-65eb-4a90-9764-86b3661aaa77",
        "tileMatrix": Array [
          Object {
            "identifier": "0",
            "matrixHeight": 4,
            "matrixWidth": 8,
            "scaleDenominator": 1,
            "tileHeight": 720,
            "tileWidth": 720,
            "topLeftCorner": Array [
              0,
              0,
            ],
            "type": "TileMatrixType",
          },
        ],
        "title": "GeoVisio tile matrix for picture 0005086d-65eb-4a90-9764-86b3661aaa77",
        "type": "TileMatrixSetType",
      },
    },
    "view:azimuth": 0,
  },
  "sequence": Object {
    "id": "bb129602-5ac1-4512-bf67-9ec1fa23033f",
    "nextPic": "ccf5f468-38b8-4b42-8d79-b97f5391e0a3",
    "prevPic": "127d9282-d082-46ca-bc09-4b0d8569dd2c",
  },
  "sphereCorrection": Object {},
}
`;

exports[`_getNodeFromAPI works with nav filter 1`] = `
Object {
  "caption": Object {
    "date": Object {
      "toLocaleDateString": [Function],
    },
    "license": "proprietary",
    "producer": "User",
  },
  "gps": Array [
    2.1822941,
    48.9810589,
  ],
  "horizontalFov": 360,
  "id": "0005086d-65eb-4a90-9764-86b3661aaa77",
  "links": Array [],
  "panorama": Object {
    "basePanoData": [Function],
    "baseUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/sd.webp",
    "cols": 8,
    "hdUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/hd.jpg",
    "origBaseUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/sd.webp",
    "rows": 4,
    "thumbUrl": "https://geovisio.fr/api/pictures/0005086d-65eb-4a90-9764-86b3661aaa77/thumb.jpg",
    "tileUrl": [Function],
    "width": 5760,
  },
  "properties": Object {
    "datetime": "2022-05-03T13:28:56+00:00",
    "license": "proprietary",
    "pers:interior_orientation": Object {
      "camera_manufacturer": "GoPro",
      "camera_model": "Max",
      "field_of_view": 360,
      "focal_length": 3,
    },
    "tiles:tile_matrix_sets": Object {
      "geovisio": Object {
        "identifier": "geovisio-0005086d-65eb-4a90-9764-86b3661aaa77",
        "tileMatrix": Array [
          Object {
            "identifier": "0",
            "matrixHeight": 4,
            "matrixWidth": 8,
            "scaleDenominator": 1,
            "tileHeight": 720,
            "tileWidth": 720,
            "topLeftCorner": Array [
              0,
              0,
            ],
            "type": "TileMatrixType",
          },
        ],
        "title": "GeoVisio tile matrix for picture 0005086d-65eb-4a90-9764-86b3661aaa77",
        "type": "TileMatrixSetType",
      },
    },
    "view:azimuth": 0,
  },
  "sequence": Object {
    "id": "bb129602-5ac1-4512-bf67-9ec1fa23033f",
    "nextPic": "ccf5f468-38b8-4b42-8d79-b97f5391e0a3",
    "prevPic": "127d9282-d082-46ca-bc09-4b0d8569dd2c",
  },
  "sphereCorrection": Object {},
}
`;

exports[`clearPictureMetadataCache works when a pic is selected 1`] = `
Array [
  Array [
    "seq",
    "pic",
  ],
]
`;

exports[`setTransitionDuration works 1`] = `
Array [
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
]
`;
